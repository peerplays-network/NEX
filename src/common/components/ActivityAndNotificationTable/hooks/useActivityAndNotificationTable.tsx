import counterpart from "counterpart";
import { uniq } from "lodash";
import { useCallback, useEffect, useState } from "react";

import { useActivity, useFormDate } from "../../../hooks";
import { useViewportContext } from "../../../providers";
import { ActivityRow } from "../../../types";
import {
  ActivityAndNotificationColumns,
  ActivityAndNotificationType,
} from "../components";

import {
  UseActivityAndNotificationResult,
  UseActivityAndNotificationTableArgs,
} from "./useActivityAndNotificationTable.types";

export function useActivityAndNotificationTable({
  userName,
  isWalletActivityTable = false,
  isNotificationTab,
  notifications,
  markTheNotificationAsReadOrUnread,
}: UseActivityAndNotificationTableArgs): UseActivityAndNotificationResult {
  const [activitiesAndNotificationsRows, _setActivitiesAndNotificationsRows] =
    useState<ActivityRow[]>([]);
  const [activityAndNotificationColumns, setActivityAndNotificationColumns] =
    useState<ActivityAndNotificationType[]>([]);
  const [searchDataSource, setSearchDataSource] = useState<ActivityRow[]>([]);
  const [loading, setLoading] = useState<boolean>(true);
  const { getActivitiesRows } = useActivity();
  const { sm } = useViewportContext();
  const { convertUTCDateToLocalDate } = useFormDate();
  const columns = ActivityAndNotificationColumns(
    isNotificationTab,
    markTheNotificationAsReadOrUnread
  );

  const formDate = useCallback(
    (
      date: string | number | Date,
      pattern = ["day", "month", "date", "year"]
    ): string => {
      const localDate = convertUTCDateToLocalDate(new Date(date));
      const newDate = String(localDate).split(" ");
      const dateObj: {
        [segment: string]: string;
      } = {
        day: newDate[0] + ",",
        date: newDate[2],
        month: newDate[1],
        year: newDate[3],
        time: newDate[4],
      };
      if (sm) {
        return (
          pattern.map((el: string) => dateObj[el]).join(" ") +
          " | " +
          dateObj.time
        );
      }

      return `${dateObj.year}-${
        localDate.getMonth() + 1
      }-${localDate.getDate()} ${dateObj.time}`;
    },
    [sm]
  );

  const setActivitiesAndNotificationsRows = useCallback(async () => {
    try {
      setLoading(true);
      let rowsDataSource, uniqTypes: string[];
      const activityRows = await getActivitiesRows(
        userName as string,
        isWalletActivityTable
      );

      if (isNotificationTab) {
        rowsDataSource = notifications.map(({ activity, unread }) => {
          return {
            ...activity,
            status: unread,
          } as ActivityRow;
        });
      } else {
        rowsDataSource = activityRows.map((activityRow) => {
          return {
            ...activityRow,
            time: formDate(activityRow.time),
          } as ActivityRow;
        });
        const allTypes = rowsDataSource.map((activity) => activity.type);
        uniqTypes = uniq(allTypes);
      }

      const updatedColumns = columns.map((column) => {
        if (column.key === "type") {
          column.filters = uniqTypes.map((type: string) => {
            return {
              text: counterpart.translate(`transaction.trxTypes.${type}.title`),
              value: type,
            };
          });
        }
        return { ...column };
      });

      setActivityAndNotificationColumns(updatedColumns);
      _setActivitiesAndNotificationsRows(rowsDataSource);
      setSearchDataSource(rowsDataSource);
      setLoading(false);
    } catch (e) {
      setLoading(false);
      console.log(e);
    }
  }, [
    setLoading,
    _setActivitiesAndNotificationsRows,
    isWalletActivityTable,
    userName,
    isNotificationTab,
    columns,
  ]);

  useEffect(() => {
    setActivitiesAndNotificationsRows();
  }, [userName, notifications]);

  return {
    activitiesAndNotificationsRows,
    loading,
    activityAndNotificationColumns,
    searchDataSource,
    setSearchDataSource,
  };
}
