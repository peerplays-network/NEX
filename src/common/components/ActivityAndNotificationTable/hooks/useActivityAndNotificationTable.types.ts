import { Dispatch, SetStateAction } from "react";

import { ActivityRow, Notification } from "../../../types";
import { ActivityAndNotificationType } from "../components";

export type UseActivityAndNotificationResult = {
  activitiesAndNotificationsRows: ActivityRow[];
  loading: boolean;
  activityAndNotificationColumns: ActivityAndNotificationType[];
  searchDataSource: ActivityRow[];
  setSearchDataSource: Dispatch<SetStateAction<ActivityRow[]>>;
};

export type UseActivityAndNotificationTableArgs = {
  userName?: string;
  isWalletActivityTable?: boolean;
  isNotificationTab: boolean;
  notifications: Notification[];
  markTheNotificationAsReadOrUnread: (id: string, unread: boolean) => void;
};
