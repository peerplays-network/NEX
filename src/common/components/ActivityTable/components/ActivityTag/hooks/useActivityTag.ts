import { TypeMap, UseActivityTag } from "./useActifityTag.types";

export function useActivityTag(): UseActivityTag {
  const getTypeString = (type: string): string => {
    const typeMap = {
      account_create: "Account Creation",
      limit_order_create: "Create Order",
      limit_order_cancel: "Cancel Order",
      fill_order: "Order Filled",
      account_update: "Account Updated",
      account_upgrade: "Account Upgraded",
      worker_create: "Worker Created",
      proposal_create: "Proposal Create",
      balance_claim: "Balance claim",
      transfer: "Transfer",
      asset_fund_fee_pool: "Fund Asset Fee Pool",
      account_whitelist: "Account Whitelist",
      asset_create: "Asset Create",
      asset_issue: "Issue Asset",
      asset_update: "Update Asset",
      asset_claim_pool: "Claim asset fee pool",
      asset_update_issuer: "Update asset issuer",
      asset_update_feed_producers: "Update asset feed producers",
    };
    return typeMap[type as keyof TypeMap];
  };
  return { getTypeString };
}
