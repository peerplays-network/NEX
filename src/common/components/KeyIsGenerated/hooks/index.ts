export { useCopyKey } from "./useCopyKey";
export { useKeyIsGenerated } from "./useKeyIsGenerated";
