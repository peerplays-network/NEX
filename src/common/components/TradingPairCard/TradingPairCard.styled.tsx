import { styled } from "../../../ui/src";
import { breakpoint } from "../../../ui/src/breakpoints";
import { colors } from "../../../ui/src/colors";

export const Card = styled.div`
  height: 100px;
  /* UI Properties */
  background: ${colors.white} 0% 0% no-repeat padding-box;
  border: 1px solid ${colors.borderColorBase};
  border-radius: 4px;
  cursor: pointer;
  transition: all 0.3s ease-out;

  &:hover {
    border: 1px solid ${colors.primaryColor};
    background: transparent
      linear-gradient(
        180deg,
        ${(props) =>
          props.theme
            ? props.theme.backgroundColorCode
            : props.theme.backgroundColorCode},
        ${colors.white} 100%
      )
      0% 0% no-repeat padding-box;
    border-radius: 4px;
  }
`;
export const ContentHeader = styled.div`
  display: flex;
  justify-content: space-between;
`;
export const TradingPair = styled.p`
  text-align: left;
  color: ${colors.textColorSecondary};
  opacity: 1;
  font-size: 14px;
  margin: 10px;
`;
export const PercentChange = styled.p`
  color: ${(props) =>
    props.theme
      ? props.theme.percentChangeColor
      : props.theme.percentChangeColor};
  font-size: 12px;
  margin: 10px;
  ${breakpoint.xs} {
    font-size: 14px;
  }
`;
export const Price = styled.p`
  text-align: left;
  color: ${colors.textColor};
  font-size: 28px;
  margin-left: 10px;
  display: ${(props) =>
    props.theme ? props.theme.display : props.theme.display};
`;
export const Volume = styled.p`
  text-align: left;
  color: ${colors.textColor};
  font-size: 26px;
  margin-left: 10px;
  ${breakpoint.xs} {
    font-size: 28px;
  }
`;
