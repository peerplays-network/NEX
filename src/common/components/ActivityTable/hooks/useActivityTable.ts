import { useCallback, useEffect, useState } from "react";

import { breakpoints } from "../../../../ui/src/breakpoints";
import { useAsset, useHistory } from "../../../hooks";
import { usePeerplaysApiContext } from "../../PeerplaysApiProvider";
import { useUserContext } from "../../UserProvider";
import { useViewportContext } from "../../ViewportProvider";

import { ActivityRow, UseActivityTable } from "./useActivityTable.types";

export function useActivityTable(): UseActivityTable {
  const [tableActivity, _setTableActivity] = useState([]);
  const [loading, setLoading] = useState<boolean>(false);
  const { dbApi } = usePeerplaysApiContext();
  const { id } = useUserContext();
  const { width } = useViewportContext();
  const { getAssetById, setPrecision } = useAsset();
  const { getHistoryById, getOperationInfo } = useHistory();

  useEffect(() => {
    setTableActivity();
  }, [id]);

  const formDate = (
    date: string | number | Date,
    pattern = ["day", "month", "date", "year"]
  ): string => {
    const newDate = String(new Date(date)).split(" ");
    const dateObj = {
      day: newDate[0] + ",",
      date: newDate[2],
      month: newDate[1],
      year: newDate[3],
      time: newDate[4],
    };
    if (width > breakpoints.sm) return String(date).replace("T", " ");
    return pattern.map((el) => dateObj[el]).join(" ") + " | "  + dateObj.time;
  };

  const formActivityRow = useCallback(
    async (activity): Promise<ActivityRow> => {
      const fee = activity.op[1].fee;
      const time = await dbApi("get_block_header", [activity.block_num]).then(
        (block: { timestamp: any }) => formDate(block.timestamp)
      );
      const info = await getOperationInfo(id, activity);
      const feeAsset = await getAssetById(fee.asset_id);
      return {
        key: activity.id,
        time,
        price: info.type,
        info: info.infoString,
        id: activity.id,
        fee: `${setPrecision(false, fee.amount, feeAsset.precision)} ${
          feeAsset.symbol
        }`,
      };
    },
    []
  );

  const setTableActivity = useCallback(async () => {
    let history = await getHistoryById(id);
    history = history.filter(
      (el: { op: number[] }) =>
        (el.op[0] >= 0 && el.op[0] <= 8) ||
        el.op[0] === 34 ||
        el.op[0] === 10 ||
        el.op[0] === 11 ||
        el.op[0] === 13 ||
        el.op[0] === 14 ||
        el.op[0] === 16
    );
    try {
      setLoading(true);
      const activityRows = await Promise.all(history.map(formActivityRow));
      _setTableActivity(activityRows);
      setLoading(false);
    } catch (e) {
      setLoading(false);
      console.log(e);
    }
  }, [tableActivity, formActivityRow, _setTableActivity, setLoading, id]);

  return { tableActivity, loading };
}
