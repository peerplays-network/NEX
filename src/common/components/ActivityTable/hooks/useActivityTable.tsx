import { useCallback, useEffect, useState } from "react";

import { useActivity } from "../../../hooks";

import {
  ActivityRow,
  UseActivityTableArgs,
  UseActivityTableResult,
} from "./useActivityTable.types";

export function useActivityTable({
  userName,
  isWalletActivityTable = false,
}: UseActivityTableArgs): UseActivityTableResult {
  const [activitiesRows, _setActivitiesRows] = useState<ActivityRow[]>([]);
  const [loading, setLoading] = useState<boolean>(true);

  const { getActivitiesRows } = useActivity();

  const setActivitiesRows = useCallback(async () => {
    try {
      setLoading(true);
      const activityRows = await getActivitiesRows(
        userName as string,
        isWalletActivityTable
      );
      _setActivitiesRows(activityRows);
      setLoading(false);
    } catch (e) {
      setLoading(false);
      console.log(e);
    }
  }, [setLoading, _setActivitiesRows, isWalletActivityTable, userName]);

  useEffect(() => {
    setActivitiesRows();
  }, [userName]);

  return { activitiesRows, loading };
}
