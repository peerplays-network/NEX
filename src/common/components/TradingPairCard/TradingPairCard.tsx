import { useRouter } from "next/router";
import React from "react";

import { useUpdateExchanges } from "../../hooks";

import * as Styled from "./TradingPairCard.styled";
import { useTradingPairStyles } from "./hooks";

type PairProps = {
  tradingPair: string;
  price: string;
  percentChange: string;
  volume: string;
};

export const TradingPairCard = ({
  tradingPair,
  price,
  percentChange,
  volume,
}: PairProps): JSX.Element => {
  const router = useRouter();
  const { updateExchanges } = useUpdateExchanges();

  const {
    handleMouseHover,
    handleMouseOut,
    changeBackgroundColor,
    showChangeAndVolume,
    positiveTheme,
    negativeTheme,
  } = useTradingPairStyles(percentChange);

  return (
    <div
      onClick={() => {
        const activePair = tradingPair.replace("/", "_");
        updateExchanges(activePair);
        router.push(`/market/${activePair}`);
      }}
    >
      <Styled.Card
        className="trading-card"
        onMouseEnter={handleMouseHover}
        onMouseLeave={handleMouseOut}
        theme={changeBackgroundColor ? positiveTheme : negativeTheme}
      >
        <Styled.ContentHeader>
          <Styled.TradingPair>{tradingPair}</Styled.TradingPair>
          {showChangeAndVolume && (
            <Styled.PercentChange
              theme={changeBackgroundColor ? positiveTheme : negativeTheme}
            >
              {changeBackgroundColor ? "+" : ""}
              {percentChange}
            </Styled.PercentChange>
          )}
        </Styled.ContentHeader>
        <Styled.Price
          theme={showChangeAndVolume ? positiveTheme : negativeTheme}
        >
          {price}
        </Styled.Price>
        {showChangeAndVolume && <Styled.Volume>{volume}</Styled.Volume>}
      </Styled.Card>
    </div>
  );
};
