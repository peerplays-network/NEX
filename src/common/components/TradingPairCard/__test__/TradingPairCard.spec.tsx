import { fireEvent, render } from "@testing-library/react";
import { useRouter } from "next/router";
import React from "react";

import { useUpdateExchanges } from "../../../hooks";
import { TradingPairCard } from "../TradingPairCard";

jest.mock("next/router", () => ({
  useRouter: jest.fn().mockReturnValue({
    push: jest.fn(),
  }),
}));

jest.mock("../../../../common/hooks", () => ({
  useUpdateExchanges: jest.fn().mockReturnValue({
    updateExchanges: jest.fn(),
  }),
}));

describe("TradingPairCard", () => {
  const tradingPair = "TEST/BTC";
  const price = "10,000";
  const percentChange = "10";
  const volume = "100,000";

  it("should renders the TradingPairCard component correctly", () => {
    const { getByText, asFragment } = render(
      <TradingPairCard
        tradingPair={tradingPair}
        price={price}
        percentChange={percentChange}
        volume={volume}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByText(tradingPair)).toBeInTheDocument();
    expect(getByText(price)).toBeInTheDocument();
  });

  it("should trigger the correct actions when clicking on the div element", () => {
    const router = { push: jest.fn() };
    (useRouter as jest.Mock).mockReturnValue(router);
    const { getByText } = render(
      <TradingPairCard
        tradingPair={tradingPair}
        price={price}
        percentChange={percentChange}
        volume={volume}
      />
    );

    const divElement = getByText(tradingPair);
    fireEvent.click(divElement);

    expect(useUpdateExchanges().updateExchanges).toHaveBeenCalledWith(
      "TEST_BTC"
    );
    expect(useRouter().push).toHaveBeenCalledWith(`/market/TEST_BTC`);
  });
});
