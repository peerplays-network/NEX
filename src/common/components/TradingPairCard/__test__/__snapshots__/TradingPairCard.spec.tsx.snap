// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TradingPairCard should renders the TradingPairCard component correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-bcXHqe OTLsA trading-card"
    >
      <div
        class="sc-gswNZR jGxrCZ"
      >
        <p
          class="sc-dkrFOg hjyPTZ"
        >
          TEST/BTC
        </p>
      </div>
      <p
        class="sc-eDvSVe iBBoZQ"
      >
        10,000
      </p>
    </div>
  </div>
</DocumentFragment>
`;
